### PR TITLE
Dualpad analog switch

### PR DESCRIPTION
**Pull request by ABOSTM**
**Summary**

* Group system IP clock enable in separate function called at init
* Manage dualpad analog switch

### CODE REVIEW

File: libraries/SrcWrapper/src/stm32/pinmap.c [hunk 1]
```diff
@@ -43,6 +65,40 @@ const uint32_t pin_map_ll[16] = {
...
+/**
+ * Configure Analog dualpad switch if necessary
+ */
+void configure_dualpad_switch(PinName pin, int function)
+{
+  PinAnalogSwitch *AnalogSwitch = (PinAnalogSwitch *) PinMapAnalogSwitch;
+
+  /* Read through PinMapAnalogSwitch array */
+  while (AnalogSwitch->pin != NC) {
+    /* Check whether pin is or is associated to dualpad Anaolg Input */
```
Comment (fpistm) [inline libraries/SrcWrapper/src/stm32/pinmap.c]:
```suggestion
    /* Check whether pin is or is associated to dualpad Analog Input */
```

File: cores/arduino/stm32/PinNames.h [hunk 2]
```diff
@@ -32,6 +32,8 @@ extern "C" {
 // PinName mask
 #define PNAME_MASK                  0xFF
 
+#define IS_STM_PIN_PDUAL(pin_name) ((pin_name & PDUAL) == PDUAL)
+
```
Comment (fpistm) [inline cores/arduino/stm32/PinNames.h]: Not used
```suggestion
```

File: libraries/SrcWrapper/src/stm32/pinmap.c [hunk 3]
```diff
@@ -43,6 +65,40 @@ const uint32_t pin_map_ll[16] = {
...
+      if (((function & STM_MODE_ANALOG) != STM_MODE_ANALOG) && ((pin & PDUAL) == PDUAL)) {
+        /**
+         * We don't configure an analog function but the pin is an analog pad
+         * (Pxy_C, ANA0 ...) In this cases Analog switch should be closed
+         */
+        LL_SYSCFG_CloseAnalogSwitch(AnalogSwitch->LL_AnalogSwitch);
+        return ;
+      } else {
+        /**
+         * Either we configure an anlog function,
```
Comment (fpistm) [inline libraries/SrcWrapper/src/stm32/pinmap.c]:
```suggestion
         * Either we configure an analog function,
```

File: libraries/SrcWrapper/src/stm32/hw_config.c [hunk 4]
```diff
@@ -19,13 +19,37 @@
 extern "C" {
 #endif
 
+/**
+  * @brief  This function enables clocks for some system IP
+  * @param  None
+  * @retval None
+  */
+void hw_config_system_IP_clock_init(void)
```
Comment (fpistm) [inline libraries/SrcWrapper/src/stm32/hw_config.c]: Wonder if this function should not be moved to clock.c ?

**Review by fpistm — CHANGES_REQUESTED** [review]

File: libraries/SrcWrapper/src/stm32/pinmap.c (hunk 1 above)
Comment (ABOSTM) [reply libraries/SrcWrapper/src/stm32/pinmap.c]: Done

File: cores/arduino/stm32/PinNames.h (hunk 2 above)
Comment (ABOSTM) [reply cores/arduino/stm32/PinNames.h]: Removed

File: libraries/SrcWrapper/src/stm32/pinmap.c (hunk 3 above)
Comment (ABOSTM) [reply libraries/SrcWrapper/src/stm32/pinmap.c]: Done

File: libraries/SrcWrapper/src/stm32/hw_config.c (hunk 4 above)
Comment (ABOSTM) [reply libraries/SrcWrapper/src/stm32/hw_config.c]: Moved and renamed

**Review by fpistm — APPROVED** [review]

**Comment by fpistm** [comment]
Integrated to official PR.

Thanks

**Pull request closed without merging**